composite/token-allocation/src/adapter.ts

```diff
@@ -35,19 +36,36 @@ export const marketCapTotalValue = (
     .toNumber()
 }
 
-const inputParams = {
-  allocations: true,
-  quote: false,
-}
-
-const toValidAllocations = (allocations: TokenAllocations): TokenAllocations => {
-  if (!allocations.every((t) => !!t.symbol))
-    throw new AdapterError({ message: `Symbol not available for all tokens.`, statusCode: 400 })
-  return allocations.map((t) => ({
-    symbol: t.symbol.toUpperCase(),
-    decimals: t.decimals || DEFAULT_TOKEN_DECIMALS,
-    balance: t.balance || DEFAULT_TOKEN_BALANCE * 10 ** (t.decimals || DEFAULT_TOKEN_DECIMALS),
-  }))
+const toValidAllocations = (allocations: any[]): TokenAllocations => {
+  const _toValidSymbol = (symbol: string) => {
+    if (!symbol)
+      throw new AdapterError({ message: `Symbol not available for all tokens.`, statusCode: 400 })
+    return symbol.toUpperCase()
+  }
+  const _toValidDecimals = (decimals: number | undefined) => {
+    if (decimals === undefined) return DEFAULT_TOKEN_DECIMALS
+    return Number.isInteger(decimals) && decimals >= 0 ? decimals : DEFAULT_TOKEN_DECIMALS
+  }
+  const _toValidBalance = (balance: number | string | undefined, decimals: number) => {
+    if (!balance) return DEFAULT_TOKEN_BALANCE * 10 ** decimals
+    let BNbalance
+    try {
+      BNbalance = BigNumber.from(balance)
```

Comment:
Converting balance into BN will make sure it doesn't receive any floating number/any other kind of value. The block looks kind of ugly, any ideas are welcome